# Engenharia de Machine Learning - XP Educação - Desafio 3 - nov/2022


## Objetivos
Exercitar os seguintes conceitos trabalhados no Módulo:
- Neste desafio, vamos trabalhar os conceitos vistos sobre métricas de desempenho, técnicas
de validação e sintonia de hiperparâmetros.
- Também vamos olhar para o fluxo completo de seleção de um modelo de aprendizado.

Dataset: https://www.openml.org/search?type=data&sort=runs&id=1480&status=active

## Carregando as bibliotecas e o dataset

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import accuracy_score

In [13]:
# carregando o dataset
df = pd.read_csv("ilpd.csv", index_col='id') # Temos uma coluna 'id' e vamos usá-la como index do dataframe
df.head()

V1      V2    V3   V4   V5  V6   V7   V8   V9   V10  Class
id                                                            
1   65  Female   0.7  0.1  187  16   18  6.8  3.3  0.90      1
2   62    Male  10.9  5.5  699  64  100  7.5  3.2  0.74      1
3   62    Male   7.3  4.1  490  60   68  7.0  3.3  0.89      1
4   58    Male   1.0  0.4  182  14   20  6.8  3.4  1.00      1
5   72    Male   3.9  2.0  195  27   59  7.3  2.4  0.40      1

In [14]:
df.tail()

V1    V2   V3   V4   V5  V6  V7   V8   V9   V10  Class
id                                                         
579  60  Male  0.5  0.1  500  20  34  5.9  1.6  0.37      2
580  40  Male  0.6  0.1   98  35  31  6.0  3.2  1.10      1
581  52  Male  0.8  0.2  245  48  49  6.4  3.2  1.00      1
582  31  Male  1.3  0.5  184  29  32  6.8  3.4  1.00      1
583  38  Male  1.0  0.3  216  21  24  7.3  4.4  1.50      2

## 1) O objetivo dessa base é avaliar se uma pessoa vai ser um paciente hepático ou não. A feature que determina a classe nessa base é a “V7”.


### Resposta: Falso, a feature V7 é de Sgot Aspartate Aminotransferase segundo a descrição no site e a feature Class é a que determina a classe

## 2) Sobre as informações de número de features e amostras, marque a alternativa CORRETA. 


In [16]:
df.shape

(583, 11)

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 583 entries, 1 to 583
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      583 non-null    int64  
 1   V2      583 non-null    object 
 2   V3      583 non-null    float64
 3   V4      583 non-null    float64
 4   V5      583 non-null    int64  
 5   V6      583 non-null    int64  
 6   V7      583 non-null    int64  
 7   V8      583 non-null    float64
 8   V9      583 non-null    float64
 9   V10     583 non-null    float64
 10  Class   583 non-null    int64  
dtypes: float64(5), int64(5), object(1)
memory usage: 54.7+ KB


### Resposta: 583 amostras e 11 features incluindo o atributo classe

## 3) Essa base conta com 3 atributos categóricos. Precisamos fazer o pré processamento desses atributos para que todos sejam numéricos.


#### Temos 2 atributos categóricos: V2 e Class apesar de ser numérico 

### Resposta: Falso, apenas 2 atributos categóricos

## 4) A base não tem dados faltantes. 


In [18]:
df.isnull().sum()

V1       0
V2       0
V3       0
V4       0
V5       0
V6       0
V7       0
V8       0
V9       0
V10      0
Class    0
dtype: int64

In [19]:
df.isna().sum()

V1       0
V2       0
V3       0
V4       0
V5       0
V6       0
V7       0
V8       0
V9       0
V10      0
Class    0
dtype: int64

### Resposta: Verdadeiro

## 5) Sobre o best_score_ encontrado durante a otimização do SVC, marque a alternativa CORRETA. 
